qd_proto/dissectors/qd.lua

```diff
@@ -0,0 +1,169 @@
+-- @return true  - if message length success read;
+--         false - if not.
+local function check_msg_len(buf, off)
+    local msg_len = utils.read_compact_int(buf, off)
+    if (msg_len == nil or msg_len < 0) then
+        return false
+    else
+        return true
+    end
+end
+
+-- Get message type.
+-- @param buf Input buffer.
+-- @param off Offset to message type.
+-- @return Message type.
+local function get_msg_type(buf, off)
+    local type = {}
+    -- Message type num.
+    type.val_uint = buf(off, 1):uint()
+    -- String representation message type.
+    type.val_str = utils.enum_val_to_str(fields.message_type, type.val_uint)
+    return type
+end
+
+-- Check receive message.
+-- @param buf Input buffer.
+-- @param off Offset in input buffer.
+-- @return  0  - this not QD message;
+--         <0  - size input buffer not enough for assemble
+--               full QD message (negative num bytes, num which are missing);
+--         >0, - full size QD message.
+function qd.check_msg(buf, off)
+    local result = check_input_buf_len(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(), "Bad input buffer length.")
+        return 0
+    end
+
+    local result = check_input_buf_cut_off(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(),
+                 "Captured packet was shorter than original, can't reassemble.")
+        return 0
+    end
+
+    local result = check_compact_len(buf, off)
+    if (result == false) then
+        dbg.info(dbg.file(), dbg.line(),
+                 "Not enough buffer length for get compact_len.")
+        return -DESEGMENT_ONE_MORE_SEGMENT
+    end
+
+    local result = check_msg_len(buf, off)
+    if (result == false) then
+        dbg.warn(dbg.file(), dbg.line(), "Bad message length.")
+        return 0
+    end
+
+    local msg_len, msg_len_size = utils.read_compact_int(buf, off)
+    local buf_len = buf:len() - off
+    local full_msg_len = msg_len_size + msg_len
+    if (full_msg_len > buf_len) then
+        dbg.info(dbg.file(), dbg.line(), "Need more data for build message.")
+        return -(full_msg_len - buf_len)
+    end
+
+    return full_msg_len
+end
+
+-- Dissect input message.
+-- @param proto Protocol object.
+-- @param tvb Input buffer.
+-- @param off Offset in input buffer
+-- @param pinfo Package info.
+-- @param tree Tree for display fields in Wireshark.
+-- @return  0  - this not QD message;
+--         <0  - size input buffer not enough for assemble
+--               full QD message (negative num bytes, num which are missing);
+--         >0, - full size QD message, message data (may be nil) and subtree.
+function qd.dissect(proto, tvb, off, pinfo, tree)
+    -- Check input buf.
+    local full_msg_len = qd.check_msg(tvb, off)
+    if (full_msg_len <= 0) then return full_msg_len end
+
+    -- Get message length.
+    local len_off = off
+    local len, len_size = utils.read_compact_int(tvb, off)
```

Comment:
There is no handling of the case when `read_compact_int` returns `nil` (`compact_len > buf:len() - off`).